tests/tas-elasticsearch/src/test/java/org/alfresco/elasticsearch/ElasticsearchReindexingTests.java

```diff
@@ -0,0 +1,263 @@
+import org.elasticsearch.client.RestHighLevelClient;
+import org.elasticsearch.index.query.QueryBuilders;
+import org.elasticsearch.index.reindex.BulkByScrollResponse;
+import org.elasticsearch.index.reindex.DeleteByQueryRequest;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.test.context.ContextConfiguration;
+import org.springframework.test.context.testng.AbstractTestNGSpringContextTests;
+import org.testcontainers.containers.GenericContainer;
+import org.testcontainers.containers.startupcheck.IndefiniteWaitOneShotStartupCheckStrategy;
+import org.testng.annotations.BeforeClass;
+import org.testng.annotations.Test;
+
+import java.io.IOException;
+import java.time.Clock;
+import java.time.ZonedDateTime;
+import java.time.format.DateTimeFormatter;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Set;
+import java.util.UUID;
+import java.util.stream.Collectors;
+
+import static org.testng.Assert.assertEquals;
+
+/**
+ * TODO Decide whether this test class needs to be in a maven submodule of its own.
+ */
+@ContextConfiguration(locations = "classpath:alfresco-elasticsearch-context.xml",
+                      initializers = AlfrescoStackInitializer.class)
+
+public class ElasticsearchReindexingTests extends AbstractTestNGSpringContextTests
+{
+    private static final Logger LOGGER = LoggerFactory.getLogger(ElasticsearchReindexingTests.class);
+
+    public static final String CUSTOM_ALFRESCO_INDEX = "custom-alfresco-index";
+
+    @Autowired
+    private ServerHealth serverHealth;
+    @Autowired
+    private DataUser dataUser;
+    @Autowired
+    private DataSite dataSite;
+    @Autowired
+    private DataContent dataContent;
+    @Autowired
+    protected RestWrapper client;
+
+    private UserModel testUser;
+
+    private SiteModel testSite;
+
+    private RestHighLevelClient elasticClient;
+
+    /**
+     * Create a user and a private site and wait for these to be indexed.
+     */
+    @BeforeClass(alwaysRun = true)
+    public void dataPreparation()
+    {
+        serverHealth.isServerReachable();
+        serverHealth.assertServerIsOnline();
+
+        Step.STEP("Create a test user and private site containing a document.");
+
+        testUser = dataUser.createRandomTestUser();
+        testSite = dataSite.usingUser(testUser).createPrivateRandomSite();
+        createDocument();
+
+        Step.STEP("create ES client");
+        elasticClient = new RestHighLevelClient(
+                RestClient.builder(new HttpHost(AlfrescoStackInitializer.elasticsearch.getContainerIpAddress(),
+                                                AlfrescoStackInitializer.elasticsearch.getFirstMappedPort(),
+                                                "http")));
+
+    }
+
+    /**
+     * This is run as the first test in the class so that we know that no other test has indexed the system documents.
```

Comment:
Do we still know that no other test has indexed the system documents?  I couldn't work out exactly when the testcontainers were created/recreated.

Reply:
yes, I forced the order of tests execution inside the testSuite XML file definition